### PR TITLE
Transactional link/unlink operations

### PR DESCRIPTION
**Pull request by kalefranz**
@soapy1 once #3508 gets merged this shouldn't look *so* monstrous.

### CODE REVIEW

File: conda/file_permissions.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            test_file = join(path, "tmp-check-write-perms")
+            try:
+                open(test_file, 'a').close()
+                os.remove(test_file)
+            except IOError:
+                return False
+        elif os.path.isfile(path):
+            try:
+                renamed_path = "%s-test-rename" % path
+                os.rename(path, renamed_path)
```
Comment (kalefranz) [inline conda/file_permissions.py]: How about instead of `os.rename` we just try to change the last-modified date?

Comment (soapy1) [inline conda/file_permissions.py]: I guess this branch didn't get the updated code, but I replaced the above with `os.utime(path, times=None)`

Comment (kalefranz) [inline conda/file_permissions.py]: oh!

Comment (kalefranz) [inline conda/file_permissions.py]: There were two commits in the old branch that weren't include.  I cherry-picked them over.

**Pull request merged**